lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,398 @@
+
+    if (url.hasAuthority) {
+      throw "pkg: URL $url must not have a host, port, username or password.";
+    }
+    if (p.isAbsolute(url.path)) {
+      throw "pkg: URL $url must not be an absolute path.";
+    }
+    if (url.path.isEmpty) {
+      throw "pkg: URL $url must not have an empty path.";
+    }
+    if (url.hasQuery || url.hasFragment) {
+      throw "pkg: URL $url must not have a query or fragment.";
+    }
+
+    var basePath = containingUrl?.scheme == 'file'
+        ? p.fromUri(containingUrl!)
+        : _entryPointPath;
+
+    var (packageName, subpath) = _packageNameAndSubpath(url.path);
+    var packageRoot = _resolvePackageRoot(packageName, basePath);
+
+    if (packageRoot == null) return null;
+    var jsonPath = p.join(packageRoot, 'package.json');
+
+    var jsonString = readFile(jsonPath);
+    Map<String, dynamic> packageManifest;
+    try {
+      packageManifest = json.decode(jsonString) as Map<String, dynamic>;
+    } catch (e) {
+      throw "'package.json' in 'pkg:$packageName' cannot be parsed.";
+    }
+
+    if (_resolvePackageExports(
+            packageRoot, subpath, packageManifest, packageName)
+        case var resolved?) {
+      if (validExtensions.contains(p.extension(resolved))) {
+        return p.toUri(resolved);
+      } else {
+        throw "The export for '${subpath ?? "root"}' in "
+            "'$packageName' resolved to '${resolved.toString()}', "
+            "which is not a '.scss', '.sass', or '.css' file.";
+      }
+    }
+    // If no subpath, attempt to resolve `sass` or `style` key in package.json,
+    // then `index` file at package root, resolved for file extensions and
+    // partials.
+    if (subpath == null) {
+      var rootPath = _resolvePackageRootValues(packageRoot, packageManifest);
+      return rootPath != null ? p.toUri(rootPath) : null;
+    }
+
+    // If there is a subpath, attempt to resolve the path relative to the
+    // package root, and resolve for file extensions and partials.
+    var relativeSubpath = p.join(packageRoot, subpath);
+    return FilesystemImporter.cwd.canonicalize(p.toUri(relativeSubpath));
+  }
+
+  @override
+  ImporterResult? load(Uri url) => FilesystemImporter.cwd.load(url);
+
+  /// Splits a [bare import
+  /// specifier](https://nodejs.org/api/esm.html#import-specifiers) `specifier`
+  /// into its package name and subpath, if one exists.
+  ///
+  /// Because this is a bare import specifier and not a path, we always use `/`
+  /// to avoid invalid values on non-Posix machines.
+  (String, String?) _packageNameAndSubpath(String specifier) {
+    var parts = p.posix.split(specifier);
+    var name = parts.removeAt(0);
+
+    if (name.startsWith('.')) {
+      throw "pkg: name $name must not start with a '.'.";
+    }
+    if (name.contains('\\')) {
+      throw "pkg: name $name must not contain a '\\'.";
+    }
+    if (name.contains('%')) {
+      throw "pkg: name $name must not contain a '%'.";
+    }
```

Comment:
> > These checks effectively add additional, unspecified restrictions to the valid form of a `pkg:` URL.
> 
> This is from the [Node Resolution Algorithm Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification) for PACKAGE_RESOLVE, so it is an implicit part of the spec already. My understanding is that these are rules that the names of valid Node packages must follow, but I can see the case for not throwing here. Given this, what do you recommend?

For package names with `.` or `\`, I think we should just consider those to be unmatchable by the Node.js package importer. Return `null` and move on, in case another importer can handle them.

For `%`, I think it's really important that `pkg:%66oo` and `pkg:foo` be identical, because that's how URLs behave. We should URL-decode the package name (probably with `p.fromUri()`) and only _then_ return `null` if it still contains a `%`.